packages/server/src/routes/api/login.ts

```diff
@@ -0,0 +1,94 @@
+import config from '../../config';
+import Router from '../../utils/Router';
+import { redirect, SubPath } from '../../utils/routeUtils';
+import { generateRedirectHtml, identityProvider, serviceProvider } from '../../utils/saml';
+import { AppContext, RouteType, SamlPostResponse } from '../../utils/types';
+import { bodyFields } from '../../utils/requestUtils';
+import { InternalServerError } from '../../utils/errors';
+import { FlowResult } from 'samlify/types/src/flow';
+import { cookieSet } from '../../utils/cookies';
+import defaultView from '../../utils/defaultView';
+
+export const router = new Router(RouteType.Api);
+
+router.public = true;
+
+// Set an error message saying that SAML is disabled.
+// @param ctx The AppContext.
+function samlNotAvailable(ctx: AppContext) {
+	ctx.status = 403;
+	ctx.body = { error: 'This server does not accept SAML authentication.' };
+}
+
+// Redirect the user to the IdP login page, if they somehow get to this URL directly.
+router.get('api/saml', async (_path: SubPath, ctx: AppContext) => {
+	if (config().saml.enabled) {
+		return await generateRedirectHtml();
+	} else {
+		return samlNotAvailable(ctx);
+	}
+});
+
+// Called when a user successfully authenticated with the IdP, and was redirected to Joplin.
+router.post('api/saml', async (_path: SubPath, ctx: AppContext) => {
+	if (config().saml.enabled) {
+		// Load SAML configuration
+		const [sp, idp] = await Promise.all([
+			serviceProvider(),
+			identityProvider(),
+		]);
+
+		// Parse the login response
+		const fields = await bodyFields<SamlPostResponse>(ctx.req);
+
+		let result: FlowResult;
+
+		try {
+			result = await sp.parseLoginResponse(idp, 'post', { body: fields });
+		} catch (error) {
+			throw new InternalServerError('Failed to parse the SAML response! Please check server configuration.', { details: { originalError: error } });
+		}
+
+		// Extract attributes from the SAML response
+		const email = result.extract.attributes['email'];
+		const displayName = result.extract.attributes['displayName'];
+
+		// Validate the attributes
+		if (typeof email !== 'string' || email === '' || typeof displayName !== 'string' || displayName === '') {
+			throw new InternalServerError('Invalid SAML response. Either the email or the display name is invalid.');
+		}
```

Comment:
`ssoLogin` should already fail if the data is invalid so this is not needed. And if `ssoLogin` doesn't throw an error on invalid data then that's where the problem is.

As a general rule we try to keep the business logic in the models and parsers and keep the route handler as small as possible (Fat Model / Skinny Controller)